designs/replicated-pv/mayastor/raid0.md

```diff
@@ -0,0 +1,145 @@
+    * [Workflow](#workflow)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Testing](#testing)
+
+## Summary
+
+This proposal adds RAID0 support to Mayastor disk pools, enabling users to combine multiple devices into single pools with aggregate capacity. The implementation leverages SPDK RAID0 functionality while preserving all existing LVS features including encryption, snapshots, and thin provisioning. This enhancement addresses current limitations where users must create separate pools per device, leading to potential sub-optimal disk space utilization across multiple small pools.
+
+## Motivation
+
+Mayastor currently requires separate disk pools per device, resulting in multiple small pools with potential sub-optimal disk space utilization. RAID0 support enables users to consolidate multiple devices into larger pools with improved capacity utilization.
+
+### Goals
+
+- Enable multi-device RAID0 pools with aggregate capacity from multiple devices
+- Preserve all existing LVS features (encryption, snapshots, thin provisioning) on RAID0 pools
+- Maintain backward compatibility with existing single-device pools
+- Provide foundation for future RAID level support (RAID1, RAID5)
+
+**Success Criteria:**
+- Users can create RAID0 pools from multiple devices
+- RAID0 pools function identically to single-device pools from user perspective
+- No regressions in existing single-device functionality
+
+### Non-Goals
+
+- RAID0 device replacement maybe future enhancement)
+- Migration from existing single-device pools to RAID0
+- Other RAID levels (RAID1, RAID5) - separate OEPs
+
+## Proposal
+
+### User Stories
+
+1. **Story 1**: As a user, I want to create large storage pools (e.g., combining 4x 2TB devices into an 8TB pool) to support applications requiring substantial storage capacity.
+
+### Workflow
+
+#### Single Device Pool (Existing)
+
+1. **Create DiskPool CRD**: Specify single device (no `poolType` needed)
+2. **Apply configuration**: Controller creates pool directly on device
+3. **Pool ready**: Single-device pool available
+
+```yaml
+apiVersion: openebs.io/v1beta3
+kind: DiskPool
+metadata:
+  name: single-pool
+spec:
+  node: node-1
+  disks:
+    - /dev/nvme0n1
+```
+
+#### RAID0 Pool (New)
+
+1. **Prepare devices**: Ensure multiple devices are available
+2. **Create DiskPool CRD**: Specify multiple devices with `poolType: "raid0"`
+3. **Controller validation**: Validates RAID0 requirements (min 2 devices)
+4. **RAID0 creation**: io-engine creates SPDK RAID0 bdev from devices
+5. **Pool creation**: LVS blobstore created on RAID0 bdev
+6. **Pool ready**: RAID0 pool available as single large pool
+
+```yaml
+apiVersion: openebs.io/v1beta3
+kind: DiskPool
+metadata:
+  name: raid0-pool
+spec:
+  node: node-1
+  poolType: "raid0"
+  disks:
+    - /dev/nvme0n1
+    - /dev/nvme1n1
+    - /dev/nvme2n1
+  raid0:                # <- optional
```

Comment:
Should this be an enum, to support future types without having to change anything?

```
enum RaidType {
     Raid0(Raid0Params)
     Raid1(Raid1Params)
     Raid5(Raid5Params)
}

Reply:
When converting to PoolArgs or gRPC I actually use an enum like this. I was not sure how idiomatic that would be for the CRD and opted for `poolType` + `raid0` namespace. I think both approaches would be extensible.

Plus the 'raid0' here is optional, as `poolType` determines the type. If `raid0` is missing we will assume the users intends to use default settings.

Reply:
Yeah I see the appeal of keeping back-compat :)
I'm ok with it like that, I'll let others comment their thoughts too @Abhinandan-Purkait @abhilashshetty04 @dsharma-dc

Reply:
Maybe let me give backwards compatible example without default raid0 settings as well:

```
apiVersion: openebs.io/v1beta3
kind: DiskPool
metadata:
  name: raid0-pool
spec:
  node: node-1
  poolType: "raid0"
  disks:
    - /dev/nvme0n1
    - /dev/nvme1n1
```

Reply:
Sounds good!